8319117: GrowableArray: Allow for custom initializer instead of copy constructor

jdksjolen opened the pull request:
Hi, 

When using at_put and at_put_grow you can provide a value which will be supplied to the constructor of each element. In other words, you can intialize each element through a copy constructor.

I suggest that we also provide a function equivalent where the function is provided a pointer to the memory to be initialized. This can be used for `NONCOPYABLE` classes, for example.

This is implemented using a SFINAE pattern because `nullptr` introduces ambiguity if you use static overload.

Currently running tier1-tier4.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [ ] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8319117](https://bugs.openjdk.org/browse/JDK-8319117): GrowableArray: Allow for custom initializer instead of copy constructor (**Enhancement** - P4)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16409/head:pull/16409` \
`$ git checkout pull/16409`

Update a local copy of the PR: \
`$ git checkout pull/16409` \
`$ git pull https://git.openjdk.org/jdk.git pull/16409/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16409`

View PR using the GUI difftool: \
`$ git pr show -t 16409`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16409.diff">https://git.openjdk.org/jdk/pull/16409.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16409#issuecomment-1785526191)

dean-long commented:
How do you see this being used?  Shouldn't the filler be part of the template type for the class, so it can be used in the ctor?  Is there really a need for each call to at_put to use a different filler function?

stefank commented:
I'd also like to get a better understand how this is intended to use. It also lacks tests for this feature. Maybe you could write a gtest show-casing how this is supposed to be used?

I'd prefer if there was a way to get rid of the SFINAE and the code duplication, if possible.

There are a lot of inconsistencies of the formatting within the patch itself and with regards to the style used in the rest of the code. Could you take an extra pass add appropriate whitespaces and newlines?

merykitty commented:
I think a more preferable approach is to do use emplace-like filling

    template <class... Args>
    E& at_grow(int i, Args... args) {
        assert(0 <= i, "negative index %d", i);
        if (i >= this->_len) {
            if (i >= this->_capacity) {
                grow(i);
            }
            for (int j = this->_len; j <= i; j++) {
                _data[j].~E();
                new (&_data[j]) E(args...);
            }
            this->_len = i + 1;
        }
        return _data[i];
    }

jdksjolen commented:
> I think a more preferable approach is to do emplace-like filling
> 
> ```
> template <class... Args>
> E& at_grow(int i, Args... args) {
>     assert(0 <= i, "negative index %d", i);
>     if (i >= this->_len) {
>         if (i >= this->_capacity) {
>             grow(i);
>         }
>         for (int j = this->_len; j <= i; j++) {
>             _data[j].~E();
>             new (&_data[j]) E(args...);
>         }
>         this->_len = i + 1;
>     }
>     return _data[i];
> }
> ```

I think you might be right. If I understand this correctly we can pick between copy construction (having `Args` be equal to `E`) and "regular" construction depending on the arguments provided?

@stefank, @dean-long. Re: tests, yes, I should add tests. The goal here is to avoid copy construction, and I chose to provide a function so that you can yourself pick how to initialize the memory. I think @merykitty's solution might be preferable to mine.

merykitty commented:
@jdksjolen Yes you are right, note that the idiom for standard C++ looks like this

    template <class... Args>
    void emplace(Args&&... args) {
        call(std::forward<Args>(args)...);
    }

And I'm not really sure what this would become without move semantics.

jdksjolen commented:
Trying out @merykitty's idea reminded me of why this solution turned out like this: There's no 'generic' placement new-operator in `AnyObj`. This can be fixed by adding in 

```c++
class AnyObj {
  //...
  operator new (size_t, void* ptr) {
    return ptr;
  }
}
```
 
tier1 and tier2 passes with the addition of this operator.

jdksjolen commented on src/hotspot/share/memory/allocation.hpp:
```diff
@@ -494,6 +494,10 @@ class AnyObj {
 #endif // ASSERT
 
  public:
+  void* operator new(size_t size, void* ptr) {
+    return ptr;
+  }
+
```
This should probably be a separate RFE.

kimbarrett replied:
I think this change should not be made.  Callers should be using global placement new.  That's what we
do everywhere else this comes up.  (It has always seemed like a bug to me that the operation "global
placement new" is syntactically an allocation function, and so subject to this kind of name lookup
collision, when it doesn't allocate.)

jdksjolen replied:
Hi,

100% agree. To be frank, I simply forgot that we had access to the global placement new through prepending `::`.

merykitty commented on src/hotspot/share/utilities/growableArray.hpp:
```diff
@@ -401,24 +401,30 @@ class GrowableArrayWithAllocator : public GrowableArrayView<E> {
+  }
+public:
+  template<typename... Args >
+  E at_grow(int i, Args... args) {
```
Should this be `Args&... args`? Also this method is returning a value instead of a reference, is it intentional?

jdksjolen replied:
>Should this be Args&... args?

Yes.

>Also this method is returning a value instead of a reference, is it intentional?

That is difficult to say. The caller can decided whether to capture the value by-reference or by-value either way.

jdksjolen replied:
> > Should this be Args&... args?
> 
> Yes.
> 

Actually, no: We can't take by reference because most of the time we're calling `at_grow` and `at_put_grow` with rvalues and not lvalues, so we need move semantics for this to work. We have to keep the arguments as copying.

merykitty replied:
I see, in that case it must be `const Args& args`. https://godbolt.org/z/aov8v38s9

jdksjolen replied:
Fixed, thanks for the help with this.

JornVernee replied:
For perfect forwarding, `Args&&...` should be used together with `std::forward` when doing the downstream call to the constructor. That will make it so rvalue arguments are forwarded as rvalues as well. The current code turns all args into lvalues.

Do we need to support NONCOPYABLE argument types as well?

jdksjolen replied:
Hi Jorn, the style guide currently doesn't allow for using perfect forwarding or move semantics in the code, so we cannot do this at the moment.

There's a PR for that: https://github.com/openjdk/jdk/pull/15386

JornVernee replied:
Ah, ok I see. Thanks for the link.

kimbarrett replied:
This is a case where using rvalues and perfect forwarding is incorrect. The
arguments are potentially used in multiple E constructor calls. Using the same
rvalue in multiple places is very bad.

JornVernee replied:
Ah, that's right. We call it in a loop, so `args...` might be used multiple times.

merykitty commented on src/hotspot/share/utilities/growableArray.hpp:
```diff
@@ -401,24 +401,30 @@ class GrowableArrayWithAllocator : public GrowableArrayView<E> {
-
-  E at_grow(int i, const E& fill = E()) {
+private:
+  static void default_fill(E* ptr) {
```
This method is unused I believe

jdksjolen replied:
I hoped I'd manage to push a fix for that before anyone saw it :-).

kimbarrett commented on src/hotspot/share/utilities/growableArray.hpp:
```diff
@@ -402,23 +402,25 @@ class GrowableArrayWithAllocator : public GrowableArrayView<E> {
       for (int j = this->_len; j <= i; j++)
-        this->_data[j] = fill;
-      this->_len = i+1;
+        new (&this->_data[j]) E(args...);
```
Use global placement new, e.g.  `::new`.  Also below, in `at_put_grow`

kimbarrett replied:
Style: Missing braces around the for-loop body.  Also below in `at_put_grow`.

merykitty replied:
I believe you also need to destroy the old object before initialise a new one.

jdksjolen commented on src/hotspot/share/utilities/growableArray.hpp:
```diff
@@ -402,23 +402,29 @@ class GrowableArrayWithAllocator : public GrowableArrayView<E> {
+        this->_data[j].~E();
+        ::new (&this->_data[j]) E(args...);
+      }
+      this->_len = i + 1;
```
Hi,

I have this PR which is meant to simply do away with `[len, cap)` being initialized: https://github.com/openjdk/jdk/pull/16418

jdksjolen commented on src/hotspot/share/utilities/growableArray.hpp:
```diff
@@ -402,23 +402,29 @@ class GrowableArrayWithAllocator : public GrowableArrayView<E> {
 
-  void at_put_grow(int i, const E& elem, const E& fill = E()) {
+  template<typename... Args>
+  void at_put_grow(int i, const E& elem, const Args&... args) {
```
Sorry, I'm not sure what you're missing here. This change enables in-place construction of the objects and the `Args...` doesn't have to be of type `E`. Is the issue with the default being removed?

kimbarrett replied:
My point is that the old code was fine.  Copy-assigning the fill argument
(either defaulted or passed in) had to work (E can't be non-copyable), because
at_put_grow requires E to be copy-assignable in order to install the new value
at index i.

jdksjolen replied:
>at_put_grow requires E to be copy-assignable in order to install the new value
at index i.

With the new code it doesn't seem to me that `at_put_grow` requires this, correct? I'd like to get rid of this requirement.

kimbarrett replied:
See line#430: `this->_data[i] = elem;`.  That requires E to be copy-assignable.

merykitty commented on src/hotspot/share/c1/c1_GraphBuilder.cpp:
```diff
@@ -178,7 +178,7 @@ BlockBegin* BlockListBuilder::make_block_at(int cur_bci, BlockBegin* predecessor
     block->init_stores_to_locals(method()->max_locals());
     _bci2block->at_put(cur_bci, block);
-    _bci2block_successors.at_put_grow(cur_bci, BlockList());
+    _bci2block_successors.at_put_grow(cur_bci, BlockList(), BlockList());
```
You can remove this change, the objects will be default constructed in-place in absence of any addition arguments.

jdksjolen commented:
Hi,

I've reworked this PR a bit. I've introduced two new methods with a `_with` suffix (any ideas for better names?). These methods allow for supplying a functor/lambda for constructing the object, similar to what was in my original PR. The original API now uses these two methods. There are also new tests, along with a toy example showing how this new API could be useful. The `at_put_grow_with` lambda takes two arguments, the latter being a boolean flag indicating whether or not this is the last element. It's perhaps not the most elegant, but I don't think having two templates would do us any good here.

kimbarrett commented:
> [...] The `at_put_grow_with` lambda takes two arguments, the latter being a boolean flag indicating whether or not this is the last element. It's perhaps not the most elegant, but I don't think having two templates would do us any good here.

I don't see the point of the "last element" flag for the _with function
argument. It is used to change what was in the old code a copy-assign into a
copy-construct, but so what? A class should be either non-copyable (neither by
constructor nor by assignment), or it should be copyable (both by constructor
and by assignment). Anything else is weird and probably broken. So avoiding
copy-assignment isn't really buying anything except API complexity.

jdksjolen commented:
> > [...] The `at_put_grow_with` lambda takes two arguments, the latter being a boolean flag indicating whether or not this is the last element. It's perhaps not the most elegant, but I don't think having two templates would do us any good here.
> 
> I don't see the point of the "last element" flag for the _with function argument. It is used to change what was in the old code a copy-assign into a copy-construct, but so what? A class should be either non-copyable (neither by constructor nor by assignment), or it should be copyable (both by constructor and by assignment). Anything else is weird and probably broken. So avoiding copy-assignment isn't really buying anything except API complexity.

Would you prefer that we instantiate the element with global placement new?

lkorinth commented on src/hotspot/share/utilities/growableArray.hpp:
```diff
@@ -402,26 +402,46 @@ class GrowableArrayWithAllocator : public GrowableArrayView<E> {
       this->_len = i+1;
     }
-    this->_data[i] = elem;
+    ::new (&this->_data[i]) E(elem);
```
I think this needs to be guarded (there might already be an element at `i`)

jdksjolen replied:
Thanks Leo, I fixed the bug and some style issues along with it.

lkorinth commented on src/hotspot/share/utilities/growableArray.hpp:
```diff
@@ -402,26 +402,57 @@ class GrowableArrayWithAllocator : public GrowableArrayView<E> {
-        this->_data[j] = fill;
-      this->_len = i+1;
+    if (i < this->_len) {
+      this->_data[i] = E(elem);
```
This is not good if we want to remove dependency on having an assignment operator. Maybe change the assignment and the return statement to a destructor call, and let the code path below initialize the value in the destructed memory?
